Sources_v3/Controllers/MessageControllerGeneric.swift

```diff
@@ -0,0 +1,201 @@
+//
+// Copyright © 2020 Stream.io Inc. All rights reserved.
+//
+
+import CoreData
+import Foundation
+
+public extension Client {
+    func messageController(messageId: MessageId) -> MessageControllerGeneric<ExtraData> {
+        .init(client: self, messageId: messageId, environment: .init())
+    }
+}
+
+public typealias MessageController = MessageControllerGeneric<DefaultDataTypes>
+
+public final class MessageControllerGeneric<ExtraData: ExtraDataTypes>: Controller, DelegateCallable {
+    public let client: Client<ExtraData>
+        
+    public let messageId: MessageId
+    
+    public var message: MessageModel<ExtraData>? {
+        state != .inactive ? messageObserver.item : nil
+    }
+    
+    private let environment: Environment
+    
+    private(set) var anyDelegate: AnyMessageControllerDelegate<ExtraData>? {
+        didSet { stateDelegate = anyDelegate }
+    }
+    
+    private lazy var messageObserver = createMessageObserver()
+        .onChange { [unowned self] change in
+            self.delegateCallback {
+                $0?.messageController(self, didChangeMessage: change)
+            }
+        }
+    
+    private lazy var messageUpdater: MessageUpdater<ExtraData> = environment.messageUpdaterBuilder(
+        client.databaseContainer,
+        client.webSocketClient,
+        client.apiClient
+    )
+
+    init(client: Client<ExtraData>, messageId: MessageId, environment: Environment) {
+        self.client = client
+        self.messageId = messageId
+        self.environment = environment
+    }
+    
+    public func startUpdating(_ completion: ((Error?) -> Void)? = nil) {
+        do {
+            try messageObserver.startObserving()
+        } catch {
+            callback { completion?(ClientError.FetchFailed()) }
+            return
+        }
+        
+        state = .localDataFetched
+        
+        callback { completion?(nil) }
+    }
+}
+
+// MARK: - Actions
+
+public extension MessageControllerGeneric {
+    func editMessage(text: String, completion: ((Error?) -> Void)? = nil) {
+        messageUpdater.editMessage(messageId: messageId, text: text) { [weak self] error in
+            self?.callback {
+                completion?(error)
+            }
+        }
+    }
```

Comment:
This will need to a little bit smarter. What if you edit a message that hasn't been sent yet? It should just change the text locally and let the message be sent.

This logic should probably happen in the updater, not in the controller, of course.